dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -261,6 +262,121 @@ bool DMFileReader::getSkippedRows(size_t & skip_rows)
     return next_pack_id < use_packs.size();
 }
 
+size_t DMFileReader::skipNextBlock()
+{
+    // Go to next available pack.
+    size_t skip;
+    if (!getSkippedRows(skip))
+        return false;
+
+    // Find the next contiguous packs will be read in next read,
+    // let next_pack_id point to the next pack of the contiguous packs.
+    // For example, if we have 10 packs, use_packs is [0, 1, 1, 0, 1, 1, 0, 0, 1, 1],
+    // and now next_pack_id is 1, then we will skip 2 packs(index 1 and 2), and next_pack_id will be 3.
+    size_t read_pack_limit = read_one_pack_every_time ? 1 : 0;
+    const std::vector<RSResult> & handle_res = pack_filter.getHandleRes();
+    RSResult expected_handle_res = handle_res[next_pack_id];
+    auto & use_packs = pack_filter.getUsePacks();
+    size_t start_pack_id = next_pack_id;
+    const auto & pack_stats = dmfile->getPackStats();
+    size_t read_rows = 0;
+    for (; next_pack_id < use_packs.size() && use_packs[next_pack_id] && read_rows < rows_threshold_per_read; ++next_pack_id)
+    {
+        if (read_pack_limit != 0 && next_pack_id - start_pack_id >= read_pack_limit)
+            break;
+        if (enable_handle_clean_read && handle_res[next_pack_id] != expected_handle_res)
+            break;
+
+        read_rows += pack_stats[next_pack_id].rows;
+        scan_context->total_dmfile_skipped_packs += 1;
+    }
+
+    scan_context->total_dmfile_skipped_rows += read_rows;
+    next_row_offset += read_rows;
+    return read_rows;
+}
+
+Block DMFileReader::readWithFilter(const IColumn::Filter & filter)
+{
+    size_t skip_rows;
+    if (!getSkippedRows(skip_rows))
+        return {};
+
+    const auto & pack_stats = dmfile->getPackStats();
+    auto & use_packs = pack_filter.getUsePacks();
+
+    size_t start_row_offset = next_row_offset;
+
+    size_t read_rows = 0;
+    size_t next_pack_id_cp = next_pack_id;
+    while (next_pack_id_cp < use_packs.size() && read_rows + pack_stats[next_pack_id_cp].rows <= filter.size())
+    {
+        const auto begin = filter.cbegin() + read_rows;
+        const auto end = filter.cbegin() + read_rows + pack_stats[next_pack_id_cp].rows;
+        use_packs[next_pack_id_cp] = use_packs[next_pack_id_cp] && std::find(begin, end, 1) != end;
+        read_rows += pack_stats[next_pack_id_cp].rows;
+        ++next_pack_id_cp;
+    }
```

Comment:
```suggestion
    const auto begin = filter.cbegin() + read_rows;
    const auto end = filter.cbegin() + read_rows + pack_stats[next_pack_id_cp].rows;
    while (next_pack_id_cp < use_packs.size() && read_rows + pack_stats[next_pack_id_cp].rows <= filter.size())
    {
        use_packs[next_pack_id_cp] = use_packs[next_pack_id_cp] && std::find(begin, end, 1) != end;
        read_rows += pack_stats[next_pack_id_cp].rows;
       
        begin  += pack_stats[next_pack_id_cp].rows;
        end += pack_stats[next_pack_id_cp].rows;
         ++next_pack_id_cp;
         end += pack_stats[next_pack_id_cp].rows;
    }
```
Not sure my suggestion is fully correct, you can double check it.

Reply:
No differ, keep it.